### PR TITLE
SAM debug: test coverage, dotnet fixes

### PR DESCRIPTION
**Pull request by justinmk3**
- test the CONFIG phase for all providers (javascript/csharp/python)
- refactor
- fix csharp codepath for `target=code` case
- merge `master` branch
- normalize all paths

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lambda/local/debugConfiguration.ts [hunk 1]
```diff
@@ -3,11 +3,27 @@
...
+import { localize } from '../../shared/utilities/vsCodeUtils'
+import { CloudFormation } from '../../shared/cloudformation/cloudformation'
+import { CloudFormationTemplateRegistry } from '../../shared/cloudformation/templateRegistry'
 
 export const DOTNET_CORE_DEBUGGER_PATH = '/tmp/lambci_debug_files/vsdbg'
+export const AWS_SAM_DEBUG_TYPE = 'aws-sam'
+export const DIRECT_INVOKE_TYPE = 'direct-invoke'
+export const TEMPLATE_TARGET_TYPE: 'template' = 'template'
+export const CODE_TARGET_TYPE: 'code' = 'code'
+export const AWS_SAM_DEBUG_TARGET_TYPES = [TEMPLATE_TARGET_TYPE, CODE_TARGET_TYPE]
```
Comment (justinmk3) [inline src/lambda/local/debugConfiguration.ts:26]: these symbols (and functions below) were moved from `src/shared/sam/debugger/awsSamDebugger.ts` to avoid circular dependencies. 

Really, some of these files should be merged together: it is not clear how `debugConfiguration.ts` differs semantically from `awsSamDebugger.ts` ... (similar comment for numerous other gratuitous demarcations in the codebase)

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 2]
```diff
@@ -90,21 +89,18 @@ export async function makeCsharpConfig(config: SamLaunchRequestArgs): Promise<Sa
...
     const baseBuildDir = await makeBuildDir()
-    const template: CloudFormation.Template = await CloudFormation.load(config.samTemplatePath)
-    const resource = await CloudFormation.getResourceFromTemplateResources({
-        templateResources: template.Resources,
-        handlerName: config.handlerName,
-    })
-    const codeUri = getCodeUri(resource, vscode.Uri.parse(config.samTemplatePath!!))
+    const template = getTemplate(config)
+    const resource = getTemplateResource(config)
+    const codeUri = getCodeRoot(config.workspaceFolder, config)
```
Comment (justinmk3) [inline src/shared/codelens/csharpCodeLensProvider.ts:94]: cannot assume that `config.samTemplatePath` is defined here; instead use `getCodeRoot()` to get the info for both target=code and target=template cases.

File: src/shared/utilities/disposableFiles.ts [hunk 3]
```diff
@@ -71,7 +71,7 @@ export class ExtensionDisposableFiles extends DisposableFiles {
 
     public static async initialize(extensionContext: vscode.ExtensionContext): Promise<void> {
         if (!!ExtensionDisposableFiles.INSTANCE) {
-            throw new Error('ExtensionDisposableFiles already initialized')
+            return
```
Comment (justinmk3) [inline src/shared/utilities/disposableFiles.ts:74]: I don't think singletons usually are so picky about potential re-init. And throwing here makes test init more difficult.

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 4]
```diff
@@ -246,30 +282,76 @@ describe('AwsSamDebugConfigurationProvider', async () => {
...
                     runtime: validRuntime,
                 },
             }
-            assert.deepStrictEqual(
-                await debugConfigProvider.resolveDebugConfiguration(undefined, debugConfig),
-                debugConfig
-            )
+            ;(c as any).configOnly = true
+            const actual = (await debugConfigProvider.resolveDebugConfiguration(folder, c))!!
+            const expected: SamLaunchRequestArgs = {
```
Comment (justinmk3) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:307]: Tests the full CONFIG phase encapsulated in `resolveDebugConfiguration()`. 

- Fast
- Maximum test coverage without actually executing `sam`

**Comment by awschristou** [comment]
Hard to see which changes are the merge from master and which changes are yours

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 5]
```diff
@@ -246,30 +286,76 @@ describe('AwsSamDebugConfigurationProvider', async () => {
...
             })
-            assert.strictEqual(resolved!.name, name)
+            // TODO: why not respect caller-chosen name?
+            assert.strictEqual(resolved!.name, 'SamLocalDebug')
         })
-        it('target=code', async () => {
-            const debugConfig = {
+
+        it('target=code: javascript', async () => {
+            const appDir = path.join(getProjectDir(), 'integrationTest-samples/js-manifest-in-root/')
```
Comment (awschristou) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:294]: This change appears to introduce component tests that access samples intended for integration tests. 

Is there a risk that the integration test samples could get changed in the future without realizing there is an impact to tests like these? Is there value in classifying the sample content?

File: src/integrationTest-samples/csharp-plain-sam-app/template.yaml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    Function:
+        Timeout: 10
+
+Resources:
+    HelloWorldFunction:
+        Type: AWS::Serverless::Function # More info about Function Resource: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction
+        Properties:
+            CodeUri: ./src/HelloWorld/
+            Handler: HelloWorld::HelloWorld.Function::FunctionHandler
+            Runtime: dotnetcore2.1
```
Comment (awschristou) [inline src/integrationTest-samples/csharp-plain-sam-app/template.yaml:17]: Is it worth making the folder something like "csharp-plain-sam-app-21", so that a net core 3.1 sample can be slotted in beside this?

File: src/integrationTest/codelens.js.test.ts [hunk 7]
```diff
@@ -15,7 +15,7 @@ const CODELENS_TEST_TIMEOUT_MILLIS = 10000
 
 const workspaceFolder = getTestWorkspaceFolder()
 
-describe('SAM Local CodeLenses (JS)', async () => {
+describe('SAM Local (JS)', async () => {
```
Comment (awschristou) [inline src/integrationTest/codelens.js.test.ts:18]: What is the motivation for this change? Seeing a test print out "SAM Local (JS)" implies it is exercising a full flow, where this test is exercising that CodeLenses appear.

File: src/lambda/local/debugConfiguration.ts [hunk 8]
```diff
@@ -92,3 +109,73 @@ export function assertTargetKind(config: SamLaunchRequestArgs, expectedTarget: '
         throw Error(`SAM debug: invalid config: ${config}`)
     }
 }
+
+export function getCodeRoot(
```
Comment (awschristou) [inline src/lambda/local/debugConfiguration.ts:113]: Are there tests for this function? I don't see them in this PR, but it looks like this function moved from one file to another, so maybe they do, and need to be moved as well.

I'm interested in tests for the following cases:
* target is template and `templateResource?.Properties?.CodeUri` contains an absolute path or a relative path

File: src/shared/cloudformation/templateRegistry.ts [hunk 9]
```diff
@@ -40,12 +40,15 @@ export class CloudFormationTemplateRegistry {
...
      * @param templateUri vscode.Uri containing the template to load in
      */
-    public async addTemplateToRegistry(templateUri: vscode.Uri): Promise<void> {
+    public async addTemplateToRegistry(templateUri: vscode.Uri, quiet?: boolean): Promise<void> {
         const pathAsString = templateUri.fsPath
         try {
             const template = await CloudFormation.load(pathAsString)
             this.templateRegistryData.set(pathAsString, template)
         } catch (e) {
+            if (!quiet) {
```
Comment (awschristou) [inline src/shared/cloudformation/templateRegistry.ts:49]: This change appears to introduce throws into existing code. Will this cause unexpected behavior?
Consider unconditionally making the log call on line 52. (also, it looks like line 52 should be passing e as a second arg to `verbose`. The logger handles formatting and error output.

Comment (justinmk3) [reply src/shared/cloudformation/templateRegistry.ts:49]: this actually restores old behavior, to make some tests happy. Might revisit later

File: src/shared/sam/debugger/awsSamDebugger.ts [hunk 10]
```diff
@@ -105,6 +115,12 @@ export class SamDebugConfigProvider implements vscode.DebugConfigurationProvider
         if (token?.isCancellationRequested) {
             return undefined
         }
+        /**
+         * XXX: Temporary magic field for testing.
+         * 1. Disables the EXECUTE phase.
+         * 2. Returns the config.
+         */
+        const configOnly = !!(config as any).configOnly
```
Comment (awschristou) [inline src/shared/sam/debugger/awsSamDebugger.ts:123]: Could this be named more blatantly, to show that it is for internal purposes, and that it is intended to not run the invoke? For example `__disableInvoke`

File: src/shared/utilities/disposableFiles.ts (hunk 3 above)
Comment (awschristou) [reply src/shared/utilities/disposableFiles.ts:74]: This change would be misleading to callers that expect this to be initialized by `extensionContext`.
Would it make more sense to make a `setInstance` method that is intended for use by tests?

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 11]
```diff
@@ -246,30 +286,76 @@ describe('AwsSamDebugConfigurationProvider', async () => {
...
-                await debugConfigProvider.resolveDebugConfiguration(undefined, debugConfig),
-                debugConfig
-            )
+            ;(c as any).configOnly = true
+            const actual = (await debugConfigProvider.resolveDebugConfiguration(folder, c))!!
+            const expected: SamLaunchRequestArgs = {
+                type: 'node', // Input "aws-sam", output "node".
+                request: 'attach', // Input "direct-invoke", output "attach".
+                runtime: 'nodejs12.x',
+                runtimeFamily: 1,
```
Comment (awschristou) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:315]: What's this?

File: src/integrationTest/codelens.js.test.ts (hunk 7 above)
Comment (justinmk3) [reply src/integrationTest/codelens.js.test.ts:18]: originally thought these were testing SAM functionality. But they are not, so I will remove "SAM" from the title and restore "CodeLens"

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 12]
```diff
@@ -246,30 +286,76 @@ describe('AwsSamDebugConfigurationProvider', async () => {
...
+                invokeTarget: {
+                    lambdaHandler: 'my.test.handler',
+                    projectRoot: 'src',
+                    target: 'code',
+                },
+                lambda: {
+                    runtime: 'nodejs12.x',
+                },
+                localRoot: path.join(appDir, 'src'), // Normalized to absolute path.
+                name: 'SamLocalDebug', // "name": "whats in a name"
```
Comment (awschristou) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:335]: Mysterious comment

File: src/integrationTest-samples/csharp-plain-sam-app/template.yaml (hunk 6 above)
Comment (justinmk3) [reply src/integrationTest-samples/csharp-plain-sam-app/template.yaml:17]: yep. lots of things can be added at low-cost yet high test-coverage, because of the separation of "config" vs "execute" phase: most important logic happens in the "config" phase, unlike legacy code which made config-related decisions throughout the codepath.

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts (hunk 5 above)
Comment (justinmk3) [reply src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:294]: I think the "integ-samples" folder should be renamed to e.g. `test-fixtures`.

If they are changed, tests need to be updated. No escaping that...

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts (hunk 4 above)
Comment (awschristou) [reply src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:311]: Why are the emitted payloads so large? Seems a much narrower amount of data is needed in order to instruct VS Code to perform the debug session.

Comment (justinmk3) [reply src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:311]: there's a lot of opportunity to reduce the payloads, but that should wait until later.

File: src/shared/utilities/disposableFiles.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/utilities/disposableFiles.ts:74]: I will change it to throw only if `extensionContext` does not match the current instance.

File: src/lambda/local/debugConfiguration.ts (hunk 8 above)
Comment (justinmk3) [reply src/lambda/local/debugConfiguration.ts:113]: created a tracking issue. This gets full coverage indirectly because it is a core part of the "config" phase tested in this PR. I think more direct coverage should not blocking merging this PR, because this PR is blocking 2 other PRs.

File: src/shared/sam/debugger/awsSamDebugger.ts (hunk 10 above)
Comment (justinmk3) [reply src/shared/sam/debugger/awsSamDebugger.ts:123]: made that change, but this can be removed pretty quickly, there's nothing stopping us except that this PR is transitional.

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts (hunk 12 above)
Comment (justinmk3) [reply src/test/shared/sam/debugger/samDebugConfigProvider.test.ts]: fixed. note that this PR is transitional, there is  low-hanging fruit that will be addressed later.

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 13]
```diff
@@ -402,10 +403,8 @@ export async function makeCoreCLRDebugConfiguration(
     config.debuggerPath = pathutil.normalize(getDebuggerPath(config.codeRoot))
     await ensureDebuggerPathExists(config.debuggerPath)
 
-    if (os.platform() === 'win32') {
-        // Coerce drive letter to uppercase. While Windows is case-insensitive, sourceFileMap is case-sensitive.
-        codeUri = codeUri.replace(DRIVE_LETTER_REGEX, match => match.toUpperCase())
```
[the remark this answers is not in the document]
Comment (justinmk3) [reply src/shared/codelens/csharpCodeLensProvider.ts:410]: > The next maintainer that changes normalize and makes it lowercase to suit their needs will break this code.

The maintainer would see broken tests then.  I can add an explicit test for `sourceFileMap` to make it clearer.

> Why wouldn't we want this code to explicitly process the output of normalize, and make an upper case drive letter?

We can, though it's covered by tests...

[the remark this answers is not in the document]
Comment (justinmk3) [reply src/shared/codelens/csharpCodeLensProvider.ts:410]: Restored this code. And added an explicit test.

File: package.json [hunk 14]
```diff
@@ -840,7 +881,7 @@
         "json-schema-to-typescript": "^8.2.0",
         "lolex": "^4.2.0",
         "marked": "^0.7.0",
-        "mocha": "^6.2.0",
+        "mocha": "^6.2.3",
```
Comment (bryceitoc9) [inline package.json:884]: Does this bump give us anything of value?

File: src/integrationTest/sam.test.ts [hunk 15]
```diff
@@ -276,7 +276,7 @@ describe.skip('SAM Integration Tests', async () => {
                     assertCodeLensReferencesSamTemplate(codeLens, samTemplatePath)
                 }).timeout(TIMEOUT)
 
-                it('invokes the Run Local CodeLens', async () => {
+                it.skip('invokes the Run Local CodeLens', async () => {
```
Comment (bryceitoc9) [inline src/integrationTest/sam.test.ts:279]: Any reason for keeping this around? Ditto for next test.

File: src/lambda/local/debugConfiguration.ts (hunk 1 above)
Comment (bryceitoc9) [reply src/lambda/local/debugConfiguration.ts:27]: `awsSamDebugger.ts` directly relates to the `aws-sam` `vscode.DebugConfigurationProvider` whereas `debugConfiguration.ts` should be tied to running the launch config generated from `awsSamDebugger.ts`.  If anything, I'd suggest renaming `awsSamDebugger.ts` to something a bit more descriptive (`awsSamDebugConfigProvider.ts`?)

File: src/lambda/local/debugConfiguration.ts [hunk 16]
```diff
@@ -92,3 +109,73 @@ export function assertTargetKind(config: SamLaunchRequestArgs, expectedTarget: '
...
+    }
+}
+
+export function getTemplate(config: AwsSamDebuggerConfiguration): CloudFormation.Template | undefined {
+    if (config.invokeTarget.target !== 'template') {
+        return undefined
+    }
+    const templateInvoke = config.invokeTarget as TemplateTargetProperties
+    const cftRegistry = CloudFormationTemplateRegistry.getRegistry()
+    const cfnTemplate = cftRegistry.getRegisteredTemplate(templateInvoke.samTemplatePath)?.template
```
Comment (bryceitoc9) [inline src/lambda/local/debugConfiguration.ts:167]: No need to assign before return.

File: src/lambda/local/debugConfiguration.ts [hunk 17]
```diff
@@ -92,3 +109,73 @@ export function assertTargetKind(config: SamLaunchRequestArgs, expectedTarget: '
...
+    return cfnTemplate
+}
+
+export function getTemplateResource(config: AwsSamDebuggerConfiguration): CloudFormation.Resource | undefined {
+    if (config.invokeTarget.target !== 'template') {
+        return undefined
+    }
+    const templateInvoke = config.invokeTarget as TemplateTargetProperties
+    const cfnTemplate = getTemplate(config)
+    const templateResource: CloudFormation.Resource | undefined = cfnTemplate?.Resources![
```
Comment (bryceitoc9) [inline src/lambda/local/debugConfiguration.ts:177]: No need to assign before return

File: src/shared/codelens/pythonCodeLensProvider.ts [hunk 18]
```diff
@@ -201,7 +203,7 @@ export async function makePythonDebugConfig(
         type: 'python',
         request: 'attach',
         runtimeFamily: RuntimeFamily.Python,
-        outFilePath: outFilePath,
+        outFilePath: pathutil.normalize(outFilePath ?? ''),
```
Comment (bryceitoc9) [inline src/shared/codelens/pythonCodeLensProvider.ts:206]: In what cases will `outFilePath` be undefined? Should that be an error case? What happens if we do end up running this with `pathutil.normalize('')` ?

File: src/shared/utilities/pathUtils.ts [hunk 19]
```diff
@@ -24,6 +36,27 @@ export function dirnameWithTrailingSlash(path: string): string {
     return dirname
 }
 
+/**
+ * Normalizes path `p`:
+ * - Replaces backslashes "\\" with "/".
+ * - Removes redundant path separators (except initial double-slash for UNC-style paths).
+ * - Uppercases drive-letter (Windows).
```
Comment (bryceitoc9) [inline src/shared/utilities/pathUtils.ts:43]: Have you tried this on Windows? I think we need lowercase drive paths (at the very least, capital drive paths in the launch configs have been breaking for me). I can do some investigation if necessary.

File: src/shared/utilities/pathUtils.ts [hunk 20]
```diff
@@ -34,3 +67,14 @@ export function getLocalRootVariants(filePath: string): string[] {
 
     return [filePath]
 }
+
+/**
+ * "c:/foo.bar" => "/foo.bar"
+ */
+export function removeDriveLetter(f: string): string {
```
Comment (bryceitoc9) [inline src/shared/utilities/pathUtils.ts:74]: What is this for? It's only used in tests; I can't think of any practical reason to strip the Windows drive letter.

File: src/shared/sam/debugger/awsSamDebugger.ts [hunk 21]
```diff
@@ -191,21 +209,50 @@ export class SamDebugConfigProvider implements vscode.DebugConfigurationProvider
...
+                    }
+                }
                 break
-            case RuntimeFamily.Python:
-                launchConfig = await this.launchPython(launchConfig)
+            case RuntimeFamily.Python: {
+                //  Make a Python launch-config from the generic config.
+                const c: PythonDebugConfiguration = await pythonDebug.makePythonDebugConfig(
+                    launchConfig,
+                    !launchConfig.noDebug,
```
Comment (bryceitoc9) [inline src/shared/sam/debugger/awsSamDebugger.ts:238]: A few things here:
* Seems weird to negate this. Can you refactor to use `launchConfig.isDebug` instead?
* Also, the `aws-sam` launch config will run EVERYTHING as a debug.
* Why do we need to specify all of these values if we're already passing `launchConfig` to the function? Can we just pull everything at that level?

File: src/integrationTest/sam.test.ts (hunk 15 above)
Comment (justinmk3) [reply src/integrationTest/sam.test.ts:279]: these tests will be restored. out of scope for this PR.

File: src/lambda/local/debugConfiguration.ts (hunk 16 above)
Comment (justinmk3) [reply src/lambda/local/debugConfiguration.ts:167]: it is useful for debugging.

File: src/shared/codelens/pythonCodeLensProvider.ts (hunk 18 above)
Comment (justinmk3) [reply src/shared/codelens/pythonCodeLensProvider.ts:206]: >  what cases will `outFilePath` be undefined?

that kind of refinement will be worked out iteratively as we refine the "config" phase.

> What happens if we do end up running this with `pathutil.normalize('')` ?

it returns empty string. covered by tests.

File: src/shared/utilities/pathUtils.ts (hunk 20 above)
Comment (justinmk3) [reply src/shared/utilities/pathUtils.ts:74]: some paths in the config don't have a drive letter. I didn't dig into why, yet. This can be revised later.

created a tracking issue.

File: src/integrationTest/sam.test.ts (hunk 15 above)
Comment (bryceitoc9) [reply src/integrationTest/sam.test.ts:279]: Why restore them? This feature branch should kill these tests.

File: src/shared/sam/debugger/awsSamDebugger.ts (hunk 21 above)
Comment (justinmk3) [reply src/shared/sam/debugger/awsSamDebugger.ts:238]: > Seems weird to negate this. Can you refactor to use `launchConfig.isDebug` instead?

goal is to eliminate `isDebug` field. `noDebug` field is a vscode-provided field, so we should use that instead of adding similar, redundant fields.

> Also, the `aws-sam` launch config will run EVERYTHING as a debug.

? We don't always attach a debugger.


>      Why do we need to specify all of these values if we're already passing `launchConfig` to the function? Can we just pull everything at that level?

The work here is iterative. There is a lot of low-hanging fruit.

File: package.json (hunk 14 above)
Comment (justinmk3) [reply package.json:884]: That change is from master: 0f7e4fe549edd

File: src/shared/utilities/pathUtils.ts (hunk 19 above)
Comment (justinmk3) [reply src/shared/utilities/pathUtils.ts:43]: either way is fine, but perhaps we should confirm this. Some code explicitly states that uppercase is needed. 

created a tracking issue.

File: src/shared/sam/debugger/awsSamDebugger.ts (hunk 21 above)
Comment (bryceitoc9) [reply src/shared/sam/debugger/awsSamDebugger.ts:238]: > 
> > Seems weird to negate this. Can you refactor to use `launchConfig.isDebug` instead?
> 
> goal is to eliminate `isDebug` field. `noDebug` field is a vscode-provided field, so we should use that instead of adding similar, redundant fields.
> 
> > Also, the `aws-sam` launch config will run EVERYTHING as a debug.
> 
> ? We don't always attach a debugger.

Forgot that VS Code has an inherent run-without-debug mode: https://code.visualstudio.com/docs/editor/debugging#_run-mode . I was saying that [our spec](https://github.com/aws/aws-toolkit-vscode/blob/master/designs/sam-debugging/local-sam-debugging.md) has nowhere to define a run-without-debug, so the only way to do it going forward will be with Ctrl+F5.

File: src/lambda/local/debugConfiguration.ts (hunk 1 above)
Comment (justinmk3) [reply src/lambda/local/debugConfiguration.ts:27]: There is plenty of shuffling to do, but it's out of scope for this PR.

File: src/integrationTest/sam.test.ts (hunk 15 above)
Comment (justinmk3) [reply src/integrationTest/sam.test.ts:279]: running `sam` is the point of these tests. that's the part that will be restored.

File: src/shared/sam/debugger/awsSamDebugger.ts (hunk 21 above)
Comment (justinmk3) [reply src/shared/sam/debugger/awsSamDebugger.ts:238]: > I was saying that [our spec](https://github.com/aws/aws-toolkit-vscode/blob/master/designs/sam-debugging/local-sam-debugging.md) has nowhere to define a run-without-debug, so the only way to do it going forward will be with Ctrl+F5.

User invokes the `Run: Start without debugging` command. That is the standard way to run a `launch.json` without debugging.

Our spec doesn't need to mention that because it's part of the editor conventions.

File: src/test/testUtil.ts [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+        name: 'test-workspace-folder',
+        index: 0,
+    }
+    return folder
+}
+
+/**
+ * Asserts that filepaths are equal, after normalizing for platform differences.
+ */
+export function assertEqualPaths(actual: string, expected: string) {
```
Comment (awschristou) [inline src/test/testUtil.ts:30]: Does this pass or fail: `assertEqualPaths("c:\foo", "d:\foo")

Comment (justinmk3) [reply src/test/testUtil.ts:30]: it passes, because driveletters are not always present (created a tracking issue to investigate that), so they are removed here before comparing.

If one wants strict equality then one would not use this function. Meanwhile we are testing more than ever before...

File: src/shared/utilities/pathUtils.ts (hunk 19 above)
Comment (bryceitoc9) [reply src/shared/utilities/pathUtils.ts:43]: Looks like this is due to how the cftRegistry is handling paths (as in: always lowercasing). Will pipe that and all getRegistry calls through the path.normalize function.

File: src/test/testUtil.ts (hunk 22 above)
Comment (awschristou) [reply src/test/testUtil.ts:30]: > If one wants strict equality then one would not use this function.

What would you name such a function, if not "assertEqualPaths"?

Comment (justinmk3) [reply src/test/testUtil.ts:30]: there is already `assert.strictEqual()` for that purpose. The lack of "strict" implies "not strict".

Comment (awschristou) [reply src/test/testUtil.ts:30]: There might be some confusion around this function in the future, but that is reasonable.

File: src/lambda/local/debugConfiguration.ts (hunk 1 above)
Comment (bryceitoc9) [reply src/lambda/local/debugConfiguration.ts:27]: FYI looks like a fair amount refactoring to `awsSamDebugger` was done in https://github.com/aws/aws-toolkit-vscode/pull/972 . Should make these distinctions more clear (https://github.com/aws/aws-toolkit-vscode/pull/971 will be refactored into those files).

Comment (justinmk3) [reply src/lambda/local/debugConfiguration.ts:27]: yes I'm in the middle of resolving that divergence from #972
